snap/quota/quota.go

```diff
@@ -153,6 +221,253 @@ func (grp *Group) SliceFileName() string {
 	return buf.String()
 }
 
+// groupQuotaStats contains information about current quotas of a group
+// and is used by getGroupQuotaStats to contain this information
+type groupQuotaStats struct {
+	memoryLimit quantity.Size
+	memoryUsed  quantity.Size
+
+	cpuLimit int
+	cpuUsed  int
+
+	threadsLimit int
+	threadsUsed  int
+
+	allowedCPUsLimit []int
+	allowedCPUsUsed  []int
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func maxq(a, b quantity.Size) quantity.Size {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func (grp *Group) getAllowedCPUs() []int {
+	if grp.CPULimit == nil || len(grp.CPULimit.AllowedCPUs) == 0 {
+		return []int{}
+	}
+	return grp.CPULimit.AllowedCPUs
+}
+
+func (grp *Group) getTotalCPUPercentage() int {
+	if grp.CPULimit == nil || grp.CPULimit.Percentage == 0 {
+		return 0
+	}
+	return max(grp.CPULimit.Count, 1) * grp.CPULimit.Percentage
+}
+
+// getGroupQuotaStats Recursively retrieve current group quotas statistics, this should just
+// be invoked on the upper parent of a group tree, and then it will gather active quotas for the
+// tree and store them in the allQuotas paramater
+func (grp *Group) getGroupQuotaStats(allQuotas map[string]*groupQuotaStats, skipGrp *Group) *groupQuotaStats {
+	limits := &groupQuotaStats{
+		memoryLimit:      grp.MemoryLimit,
+		cpuLimit:         grp.getTotalCPUPercentage(),
+		threadsLimit:     grp.TaskLimit,
+		allowedCPUsLimit: grp.getAllowedCPUs(),
+	}
+
+	for _, subGroup := range grp.subGroups {
+		// Skip the group that has requested the count, as this is called when creating subgroups
+		// and when updating quotas. And we don't want to include the current quotas for the subgroup
+		// that actually wants to update its quotas.
+		if subGroup == skipGrp {
+			continue
+		}
+
+		subGroupLimits := subGroup.getGroupQuotaStats(allQuotas, skipGrp)
+
+		// As we count up the usage of quotas across our sub-groups we must either use the actual
+		// limits of the below sub-group, or the actual usage of the sub-group. The reason we must do this
+		// is because if the sub-group doesn't have any limit set for a quota, but the sub-group has sub-groups
+		// itself that do have limits, then we must use that value instead. Hence the max* functions.
```

Comment:
I don't understand the comment and code, for a single subgroup that has a limit set, limit > used, so used will always be limit which sounds wrong

Reply:
Or maybe I'm confused about the term "used" here? I suspect the this and groupQuotaStats needs better doc to explain what their purpose is, and maybe different terminology

Reply:
I have hopefully improved the documentation now!